File: frontend/utils/marks.ts
/**
 * WAM and GPA, computed in the browser.
 *
 * The scoring tables come from `/v1/marks/reference`, which is where they are
 * pinned against Monash's own worked examples. This file is only the two
 * weighted averages plus the parsing, and it runs client-side on purpose: a
 * student's marks never leave their machine, so there is no request body with
 * somebody's failed unit in it and nothing in an access log.
 *
 * Every rule that is easy to get wrong is in `applyWam` and `applyGpa`, and
 * each one is there because a homemade calculator usually gets it wrong the
 * other way:
 *
 * - a fail counts, at its actual mark
 * - a withdrawn fail counts as zero and keeps its weight in the denominator
 * - the level weight applies to WAM only, never to GPA
 * - the level weight goes in the numerator *and* the denominator
 */

export interface MarksReference {
  grade_points: Record<string, Record<string, number>>
  grade_names: Record<string, string>
  grade_bands: { floor: number; grade: string }[]
  excluded_grades: string[]
  level_weights: { first_year: number; other: number }
  sources: Record<string, string>
  guides: Record<string, string>
}

export interface MarkEntry {
  id: number
  unitCode: string
  title?: string | null
  creditPoints: number | null
  level: number | null
  /** The mark out of 100. Null when only a grade is known. */
  mark: number | null
  /** An explicit grade code, for WN and the excluded grades. */
  grade: string | null
}

export const AUSTRALIA = 'australia'
export const MALAYSIA = 'malaysia'

/** Monash rounds the final mark, so 79.51 is an 80 and an HD. */
export function roundMark(mark: number): number {
  return mark >= 0 ? Math.floor(mark + 0.5) : Math.ceil(mark - 0.5)
}

export function gradeForMark(mark: number, ref: MarksReference): string | null {
  const rounded = roundMark(mark)
  if (rounded < 0 || rounded > 100) return null
  for (const band of ref.grade_bands) {
    if (rounded >= band.floor) return band.grade
  }
  return null
}

/** The grade an entry scores under, or null when it does not count at all. */
export function resolveGrade(entry: MarkEntry, ref: MarksReference): string | null {
  if (entry.grade) {
    const code = entry.grade.trim().toUpperCase()
    if (ref.excluded_grades.includes(code)) return null
    return code in ref.grade_names ? code : null
  }
  if (entry.mark === null || Number.isNaN(entry.mark)) return null
  return gradeForMark(entry.mark, ref)
}

function levelWeight(entry: MarkEntry, ref: MarksReference): number {
  return entry.level === 1 ? ref.level_weights.first_year : ref.level_weights.other
}

/** Entries that can actually be scored — the rest are shown but not counted. */
export function countable(entries: MarkEntry[], ref: MarksReference): MarkEntry[] {
  return entries.filter(e => e.creditPoints !== null && resolveGrade(e, ref) !== null)
}

export function applyWam(entries: MarkEntry[], ref: MarksReference): number | null {
  let top = 0
  let bottom = 0
  for (const entry of entries) {
    const grade = resolveGrade(entry, ref)
    if (grade === null || entry.creditPoints === null) continue
    const weight = entry.creditPoints * levelWeight(entry, ref)
    // A withdrawn fail has no mark and scores zero — but its weight stays in
    // the denominator, which is exactly what makes a WN so expensive.
    const mark = grade === 'WN' ? 0 : entry.mark
    if (mark === null || Number.isNaN(mark)) continue
    top += mark * weight
    bottom += weight
  }
  return bottom > 0 ? top / bottom : null
}

export function applyGpa(
  entries: MarkEntry[],
  ref: MarksReference,
  scale: string
): number | null {
  const points = ref.grade_points[scale]
  if (!points) return null
  let top = 0
  let bottom = 0
  for (const entry of entries) {
    const grade = resolveGrade(entry, ref)
    if (grade === null || entry.creditPoints === null) continue
    // No level weight here. GPA does not use it; only WAM does.
    top += points[grade]! * entry.creditPoints
    bottom += entry.creditPoints
  }
  return bottom > 0 ? top / bottom : null
}

/**
 * The same mark in every planned unit that would reach `target` WAM.
 *
 * Returned unclamped on purpose. Above 100 means the target is already out of
 * reach and below 0 means it is already secured, and both of those are more
 * useful answers than a number pinned to the end of the scale.
 */
export function markNeeded(
  done: MarkEntry[],
  plannedWeight: number,
  target: number,
  ref: MarksReference
): number | null {
  if (plannedWeight <= 0) return null
  let top = 0
  let bottom = 0
  for (const entry of done) {
    const grade = resolveGrade(entry, ref)
    if (grade === null || entry.creditPoints === null) continue
    const weight = entry.creditPoints * levelWeight(entry, ref)
    const mark = grade === 'WN' ? 0 : entry.mark
    if (mark === null || Number.isNaN(mark)) continue
    top += mark * weight
    bottom += weight
  }
  return (target * (bottom + plannedWeight) - top) / plannedWeight
}

/**
 * Pull unit codes, marks and grades out of text pasted from WES.
 *
 * Deliberately tolerant rather than clever. WES has changed its results layout
 * more than once and students paste from the transcript, the portal and a
 * spreadsheet, so this looks for the two things every one of those has on a
 * line — a unit code and a number — instead of trying to match a table shape.
 *
 * A line that has a code but no mark still comes back, so the entry appears in
 * the table with the mark blank rather than being dropped silently.
 */
const UNIT_CODE = /\b([A-Z]{2,4}\d{4})\b/
const GRADE_TOKEN = /\b(HD|NH|NP|WN|SFR|NAS|NE|WI|PGO|NPGO|WNGO|[NDCP])\b/
const NUMBER = /\b(\d{1,3}(?:\.\d+)?)\b/g

/** The credit point values the Handbook actually uses. */
const CREDIT_POINT_VALUES = [3, 6, 9, 12, 18, 24, 36, 48]
/** Grades that exist without a mark behind them. */
const MARKLESS_GRADES = ['WN', 'SFR', 'NAS', 'NE', 'WI', 'PGO', 'NPGO', 'WNGO']

export interface ParsedRow {
  unitCode: string
  mark: number | null
  grade: string | null
  creditPoints: number | null
}

export function parsePastedResults(text: string): ParsedRow[] {
  const rows: ParsedRow[] = []
  const seen = new Set<string>()

  for (const rawLine of text.split(/\r?\n/)) {
    const line = rawLine.trim()
    if (!line) continue
    const codeMatch = UNIT_CODE.exec(line.toUpperCase())
    if (!codeMatch) continue
    const unitCode = codeMatch[1]!
    if (seen.has(unitCode)) continue
    seen.add(unitCode)

    // Numbers after the code, so the digits inside the code itself are not
    // read as a mark.
    const after = line.toUpperCase().slice(codeMatch.index + unitCode.length)
    const numbers = [...after.matchAll(NUMBER)].map(m => Number(m[1]))

    const gradeMatch = GRADE_TOKEN.exec(after)
    const grade = gradeMatch ? gradeMatch[1]! : null
    // A mark is 0-100; credit points are one of the Handbook's small values, so
    // a lone 6, 12 or 24 sitting beside a mark is the credit points, not a mark.
    const marks = numbers.filter(n => n >= 0 && n <= 100)
    let creditPoints: number | null = null
    let mark: number | null = null

    if (marks.length >= 2 && CREDIT_POINT_VALUES.includes(marks[0]!)) {
      creditPoints = marks[0]!
      mark = marks[1]!
    } else if (marks.length === 1 && grade && CREDIT_POINT_VALUES.includes(marks[0]!)) {
      // One number and a grade: a withdrawn fail or an excluded grade has no
      // mark at all, so the number is the credit points. Reading it as a mark
      // of 6 would drop the unit out of the WAM entirely — and a WN has to stay
      // in the denominator, which is the whole reason it is expensive.
      creditPoints = marks[0]!
    } else if (marks.length >= 1) {
      mark = marks[marks.length - 1]!
    }

    // These grades never carry a mark, whatever else was on the line.
    if (grade && MARKLESS_GRADES.includes(grade)) mark = null

    rows.push({
      unitCode,
      mark: mark === null || Number.isNaN(mark) ? null : mark,
      grade,
      creditPoints
    })
  }
  return rows
}

File: backend/app/knowledge/marks.py
"""The rules behind WAM and GPA.

The tables here are transcribed from the two official pages this platform
already crawls - `guides/wam` and `guides/gpa` - and the tests check the
arithmetic against Monash's own worked examples from those same pages. That is
the point of keeping them in the backend rather than in the calculator's
JavaScript: a number that decides whether somebody thinks they are on track for
honours should be pinned by a test, not by whoever last edited a component.

The calculator itself runs in the browser. It fetches these tables once and does
the two weighted averages locally, so a student's marks never leave their
machine. This module is where the tables come from and where the maths is
proved; `frontend/utils/marks.ts` is the copy that runs.

Four rules that are easy to get wrong, and all four are in the official worked
examples:

* **A fail counts, at its actual mark.** MON1003 scored 40 and is in the WAM at
  40, not dropped. Most homemade calculators drop it and flatter the student.
* **A withdrawn fail counts as zero, and stays in the denominator.** MON2002 has
  no mark at all; it contributes 0 to the top and its full weighted credit
  points to the bottom.
* **First-year undergraduate units are weighted 0.5.** Not "the year you took
  it" - the level of the unit. A Level 1 unit taken in third year is still 0.5.
* **Credit points weight both.** The 0.5 must appear in the numerator *and* the
  denominator; halving only the top is the classic error that turns a 90 into
  a 45.

Malaysia is a different scale. CGPA is not GPA with different rounding - Credit
is 2.85 rather than 2.0 - and a platform serving both campuses cannot use one
table for both.
"""
from __future__ import annotations

from dataclasses import dataclass

AUSTRALIA = "australia"
MALAYSIA = "malaysia"

# From guides/gpa, "Grades and their GPA grade value" and "...CGPA grade value".
GRADE_POINTS: dict[str, dict[str, float]] = {
    AUSTRALIA: {
        "HD": 4.0,
        "D": 3.0,
        "C": 2.0,
        "P": 1.0,
        "NP": 0.7,
        "N": 0.3,
        "NH": 0.3,
        "WN": 0.0,
    },
    MALAYSIA: {
        "HD": 4.0,
        "D": 3.67,
        "C": 2.85,
        "P": 2.15,
        "NP": 1.7,
        "N": 1.15,
        "NH": 1.15,
        "WN": 0.0,
    },
}

# The code Monash prints on a transcript, and what it is called in full.
GRADE_NAMES: dict[str, str] = {
    "HD": "High distinction",
    "D": "Distinction",
    "C": "Credit",
    "P": "Pass",
    "NP": "Near pass",
    "N": "Fail",
    "NH": "Hurdle fail",
    "WN": "Withdrawn fail",
}

# Mark bands, for turning a mark into a grade. Monash rounds the final mark
# first, so 79.5 is a Distinction and 79.4 is not.
GRADE_BANDS: tuple[tuple[int, str], ...] = (
    (80, "HD"),
    (70, "D"),
    (60, "C"),
    (50, "P"),
    (45, "NP"),
    (0, "N"),
)

# Grades carrying no grade point at all: excluded from both averages, top and
# bottom. From the "Grades not included in the calculation" list on guides/gpa.
EXCLUDED_GRADES: tuple[str, ...] = (
    "SFR", "NE", "NAS", "WI", "PGO", "NPGO", "WNGO",
)

FIRST_YEAR_WEIGHT = 0.5
OTHER_YEAR_WEIGHT = 1.0


class MarkError(ValueError):
    """The entry cannot be scored - an unknown grade, or a mark out of range."""


@dataclass(frozen=True, slots=True)
class Entry:
    """One completed unit: what it was worth, and how it went."""

    unit_code: str
    credit_points: float
    # 1 for a first-year undergraduate unit. Anything else weights 1.0.
    level: int | None = None
    mark: float | None = None
    grade: str | None = None

    @property
    def level_weight(self) -> float:
        return FIRST_YEAR_WEIGHT if self.level == 1 else OTHER_YEAR_WEIGHT


def round_mark(mark: float) -> int:
    """Monash rounds the final mark, so 79.51 is 80 and an HD.

    Python's round() is banker's rounding - round(79.5) is 80 but round(78.5) is
    78 - which is not what a student sees on a transcript.
    """
    return int(mark + 0.5) if mark >= 0 else int(mark - 0.5)


def grade_for_mark(mark: float) -> str:
    rounded = round_mark(mark)
    if not 0 <= rounded <= 100:
        raise MarkError(f"a mark of {mark} is not between 0 and 100")
    for floor, grade in GRADE_BANDS:
        if rounded >= floor:
            return grade
    raise MarkError(f"no grade band for {mark}")  # pragma: no cover - unreachable


def resolve_grade(entry: Entry) -> str | None:
    """The grade this entry scores under, or None when it does not count."""
    if entry.grade:
        code = entry.grade.strip().upper()
        if code in EXCLUDED_GRADES:
            return None
        if code not in GRADE_NAMES:
            raise MarkError(f"unknown grade: {entry.grade}")
        return code
    if entry.mark is None:
        return None
    return grade_for_mark(entry.mark)


def wam(entries: list[Entry]) -> float | None:
    """Σ(mark × credit points × level weight) ÷ Σ(credit points × level weight).

    A withdrawn fail has no mark and scores 0, but its credit points stay in the
    denominator - that is what makes a WN so expensive.
    """
    top = 0.0
    bottom = 0.0
    for entry in entries:
        grade = resolve_grade(entry)
        if grade is None:
            continue
        weight = entry.credit_points * entry.level_weight
        mark = 0.0 if grade == "WN" else entry.mark
        if mark is None:
            # A grade with no mark that is not a WN - nothing to average.
            continue
        top += mark * weight
        bottom += weight
    return top / bottom if bottom else None


def gpa(entries: list[Entry], scale: str = AUSTRALIA) -> float | None:
    """Σ(grade value × credit points) ÷ Σ credit points.

    Note what is *not* here: the level weight. GPA does not use it - only WAM
    does. Applying it to both is a common and quiet mistake.
    """
    points = GRADE_POINTS[scale]
    top = 0.0
    bottom = 0.0
    for entry in entries:
        grade = resolve_grade(entry)
        if grade is None:
            continue
        top += points[grade] * entry.credit_points
        bottom += entry.credit_points
    return top / bottom if bottom else None


def mark_needed(
    entries: list[Entry], planned: list[Entry], target: float
) -> float | None:
    """The same mark in every planned unit that would reach ``target`` WAM.

    This is the question students actually ask - "what do I need this semester?"
    - and the answer is often outside 0-100, which is itself the useful answer:
    it means the target is already out of reach, or already guaranteed.

    Returns None when the planned units carry no weight at all.
    """
    done_top = 0.0
    done_bottom = 0.0
    for entry in entries:
        grade = resolve_grade(entry)
        if grade is None:
            continue
        weight = entry.credit_points * entry.level_weight
        mark = 0.0 if grade == "WN" else entry.mark
        if mark is None:
            continue
        done_top += mark * weight
        done_bottom += weight

    planned_weight = sum(e.credit_points * e.level_weight for e in planned)
    if planned_weight <= 0:
        return None
    # target = (done_top + x·planned_weight) / (done_bottom + planned_weight)
    return (target * (done_bottom + planned_weight) - done_top) / planned_weight


def reference() -> dict:
    """The tables, for the calculator to fetch once and compute against."""
    return {
        "grade_points": GRADE_POINTS,
        "grade_names": GRADE_NAMES,
        "grade_bands": [{"floor": floor, "grade": grade} for floor, grade in GRADE_BANDS],
        "excluded_grades": list(EXCLUDED_GRADES),
        "level_weights": {"first_year": FIRST_YEAR_WEIGHT, "other": OTHER_YEAR_WEIGHT},
    }
